### DISCUSSION TITLE
Supplying Initial Data

### DISCUSSION THREAD

**Discussion by alexanderson1993** [Thorium-Sim/thorium-nova#108] [Q&A]
This is more a technical question that has some user experience implications.

We want to have the first load experience for Thorium be really easy and seamless. That means a person downloads Thorium, starts it up, and within minutes has a flight going with their friends connected and playing. That also means Thorium Nova needs to ship with a default plugin that contains assets, ship definitions, a built-out universe, etc. 

What is the best way of getting that default plugin to the player. I can see two options:

- The plugin, along with all of its assets, are bundled with Thorium Nova. When Thorium Nova first starts up, it copies the plugin to the Thorium data folder.
- The plugin is stored on a web server. When Thorium Nova first starts up, it checks to see if there is any data in the Thorium folder. If there is not, it spends a bit of time downloading the plugin from the server and moves it into the Thorium data folder.

Option 2 is what Thorium Classic did, and it seemed to work fine.

Pros of Option 2: It keeps the main Thorium bundle size small, since it doesn't have to include all of the assets. It also might make the plugin development process a little simpler, keeping the plugin repo separate from the main repo.

Cons: It might slow down people getting started, since they have to wait for a second download to finish before starting their flight. Plus they might have downloaded Thorium Nova and then gone offline before starting it for the first time, which would cause an error.

I'm leaning towards Option 1, but I would love to hear other thoughts.

**Comment by alexanderson1993** [comment]
Perhaps a bit of both? I think the pros of option 1 (keeping it all bundled, mostly) outweigh the cons, but I can see reasons why Option 2 might be necessary.

For example, someone decides they want to go in and mess around with the default plugin that ships with Thorium Nova. They break something or change things beyond repair. They want a way to re-download the default plugin fresh, which we let them do with a click. 

And I already have said that I want to have a plugin registry, where people can submit and upload their own plugins for download. It makes sense this would also be the mechanism for distributing updates to the default plugin separate from full-blown Thorium Nova updates. (As a reminder, plugin updates include new ships, improved graphics, better sounds, new configurations, new and updated missions, etc)

I think making this work with Option 1 would involve having the Thorium Nova app build process download and embed the latest default plugin from the registry so folks starting out on Thorium Nova can still get up and running quickly.

**Reply by Arcadianer** [reply]
I can definitely see that working out quite nicely. It gives the players immediately something to do while also providing the possibility to update content that has been altered between the build and the play time.

**Reply by Arcadianer** [reply]
Or you could do something like they do it in some EA games.
Embed only "some" content in the actual build and have the rest stream in during playtime.
Like, while they are playing essentially a "demo version" of the game, the rest will be downloaded in the background and after the demo is finished you "reload" the plugins and the rest is available.

**Reply by alexanderson1993** [reply]
I like that suggestion. I propose we don't implement it until we see how big the full bundle is. If it's less than 100mb, I don't think it would be worthwhile, but if it's more than that, having a demo plugin bundled with the controls with a full plugin download in the background is a great idea.
